storage_v2/factory.go

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package storage_v2
+
+import (
+	"context"
+)
+
+// Factory is a general factory interface to be reused across different storage factories.
+// It lives within the OTEL collector extension component's lifecycle.
+// The Initialize and Close functions supposed to be called from the
+// OTEL component's Start and Shutdown functions.
+type Factory interface {
+	// Initialize performs internal initialization of the factory,
+	// such as opening connections to the backend store.
+	Initialize(ctx context.Context) error
```

Comment:
Cf https://github.com/open-telemetry/opentelemetry-collector/issues/9324

Reply:
I've read through the discussions and the latest approach suggested that the `Start` function shouldn't throw error anymore and uses the `TelemetrySettings.ReportStatus` to report `StatusFatalError` to halt the startup asynchronously, cmiiw.

I'm thinking that any changes to the component lifecycles should be handled by the Jaeger storage extension, we can capture the errors from the storage factory in the extension and report them instead, so the storage factory interface should not be affected by it.

Reply:
Fair enough, this will be similar to what jaeger-v1 does today. However, jaeger-v1 has a failure mode today where if the storage is not ready for any reason when the binary starts (even though the Jaeger configuration pointing to the storage is correct), Jaeger fails to start and the binary exists. That's not how a component of a distributed system technically should be operating, i.e. a db temporarily not available at start-up vs. becoming temporarily not available during normal operation should not be treated as different conditions. We don't exit the binary if we lose a connection later, so why do we kill it if we cannot establish the connection during Init()?

If we were to fix this, we need to establish a different expectation from storage implementations, and this is where async runtime status reporting instead of returning an error from Init is coming from. This https://github.com/open-telemetry/opentelemetry-collector/issues/9324#issuecomment-1915725314 has a good explanation.